File: Data Structures & Algorithms/surrounded-regions/submission-0.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        N, M = len(board),len(board[0])

        def dfs(r, c):
            if r < 0 or r >= N or c < 0 or c >= M or board[r][c] != 'O':
                return
            board[r][c] = 'T'
            dfs(r+1,c)
            dfs(r-1,c)
            dfs(r,c+1)
            dfs(r, c-1)
        
        for r in range(N):
            if board[r][0] == 'O':
                dfs(r,0)
            if board[r][M-1] == 'O':
                dfs(r,M-1)
        for c in range(M):
            if board[0][c] == 'O':
                dfs(0,c)
            if board[N-1][c] == 'O':
                dfs(N-1,c)
        for r in range(N):
            for c in range(M):
                if board[r][c] == 'T':
                    board[r][c] = 'O'
                elif board[r][c] == 'O':
                    board[r][c] = 'X'
            

File: Data Structures & Algorithms/valid-sudoku/submission-0.java
class Solution {
    record Quarter(int r, int c) {}
    ;

    public boolean isValidSudoku(char[][] board) {
        Map<Integer, Set<Integer>> rows = new HashMap<>();
        Map<Integer, Set<Integer>> cols = new HashMap<>();
        Map<Quarter, Set<Integer>> quarters = new HashMap<>();
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                int val = board[r][c];
                if (val == '.') {
                    continue;
                }
                rows.putIfAbsent(r, new HashSet<>());
                if (rows.get(r).contains(val)) {
                    return false;
                }
                rows.get(r).add(val);
                cols.putIfAbsent(c, new HashSet<>());
                if (cols.get(c).contains(val)) {
                    return false;
                }
                cols.get(c).add(val);
                Quarter q = new Quarter(r / 3, c / 3);
                quarters.putIfAbsent(q, new HashSet<>());
                if (quarters.get(q).contains(val)) {
                    return false;
                }
                quarters.get(q).add(val);
            }
        }
        return true;
    }
}
